LFS-724: As an API consumer, I can see when a resource was last updated in any of the JSON export formats

sdumitriu opened the pull request.

sdumitriu committed:
LFS-724 / LFS-778: Make forms, subjects, and vocabularies track their lastModified status
Note: this date is only updated upon checkin, so curretly it will not do anything other than copy the created date until checkin-checout is implemented

sdumitriu committed:
LFS-724 / LFS-779: Keep lastModified in .simple and .bare JSONs

sdumitriu committed:
LFS-724 / LFS-782: The data import servlet must checkout/checkin modified nodes

sdumitriu committed:
LFS-724 / LFS-782: The data import servlet must checkout/checkin modified nodes

sdumitriu committed:
[cleanup] Applied codestyle

sdumitriu committed:
LFS-724 / LFS-830: Questionnaires should be checked in at startup

marta- commented on modules/data-entry/src/main/java/ca/sickkids/ccm/lfs/DataImportServlet.java:
```diff
@@ -388,7 +379,7 @@ private Resource getOrCreateAnswer(final Resource form, final Node question)
         }
 
-        Map<String, Object> answerProperties = new LinkedHashMap<>();
+        Map<String, Object> answerProperties = new HashMap<>();
```
Why these changes?

sdumitriu replied:
Unrelated, there's a separate cleanup commit for this, but kind of needed because I need to remove one import to still fit into our checkstyle requirements. Also, fits in with the "use the least specific class needed for the task" strategy,

marta- commented on modules/data-entry/src/main/java/ca/sickkids/ccm/lfs/DataImportServlet.java:
```diff
@@ -831,8 +834,8 @@ private String findSubjectId(CSVRecord row, Node typeNode)
         String result = null;
-        String[] suffices = {"", " ID"};
-        for (String suffix: suffices) {
+        String[] suffices = { "", " ID" };
```
Out of scope, but isn't the plural `suffixes`?

sdumitriu replied:
Yes it is.

IntegralProgrammer commented on modules/data-entry/src/main/java/ca/sickkids/ccm/lfs/DataImportServlet.java:
```diff
@@ -808,10 +807,14 @@ private Node createSubject(String subjectKey, String subjectId, Node typeNode, N
-            return null;
+            LOGGER.warn("Failed to create new subject {}: {}", subjectKey, e.getMessage(), e);
+        } catch (RepositoryException e) {
+            LOGGER.warn("Failed to checking new subject {}: {}", subjectKey, e.getMessage(), e);
```
Should "checking" be changed to "check in"?

sdumitriu replied:
```suggestion
            LOGGER.warn("Failed to check in new subject {}: {}", subjectKey, e.getMessage(), e);
```

sdumitriu commented on modules/data-entry/src/main/java/ca/sickkids/ccm/lfs/DataImportServlet.java:
```diff
@@ -81,6 +80,7 @@
 )
 @SlingServletResourceTypes(resourceTypes = { "lfs/FormsHomepage" }, methods = { "POST" })
 @SlingServletName(servletName = "Data Import Servlet")
+@SuppressWarnings("checkstyle:ClassFanOutComplexity")
```
Oh, I see the checkstyle override is still here, I'll check if it still needed after removing `LinkedHashMap`.

marta- replied:
Is this checkstyle override still needed @sdumitriu ?

sdumitriu replied:
Nope, still too high.

IntegralProgrammer reviewed (dismissed):
I tested and also noticed this issue:

1. Build the `LFS-724` branch
2. Start CARDS in `dev,lfs` runMode
3. Create a new Demographics Form, enter some data, and save it
4. Open Composum and examine the `jcr:lastModified` value for the saved Form
5. Update the Demographics Form with more data, and save it.
6. Refresh Composum and observe that the `jcr:lastModified` value for this Form is not updated.

marta- commented:
> I tested and also noticed this issue:
> 
> 1. Build the `LFS-724` branch
> 2. Start CARDS in `dev,lfs` runMode
> 3. Create a new Demographics Form, enter some data, and save it
> 4. Open Composum and examine the `jcr:lastModified` value for the saved Form
> 5. Update the Demographics Form with more data, and save it.
> 6. Refresh Composum and observe that the `jcr:lastModified` value for this Form is not updated.

I think that's expected, because the frontend doesn't use the checkin/checkout mechanism yet and lastModified just defaults to thw creation timestamp. https://phenotips.atlassian.net/browse/LFS-831 and https://phenotips.atlassian.net/browse/LFS-832 are supposed to implement it.

IntegralProgrammer committed:
LFS-724 / LFS-926: Update jcr:lastModified and jcr:lastModifiedBy when updating a mix:lastModified node
Update jcr:lastModified and jcr:lastModifiedBy Form and Questionnaire properties upon check-in or check-out

IntegralProgrammer committed:
LFS-724 / LFS-831: Checkout a form when starting to edit it

IntegralProgrammer committed:
LFS-724 / LFS-831: Checkout a form when starting to edit it

Use the same useEffect() as saveData()

IntegralProgrammer committed:
LFS-724 / LFS-832: Checkin a form when closing it

IntegralProgrammer committed:
LFS-724 / LFS-833: Checkout a questionnaire when starting to edit it

IntegralProgrammer committed:
LFS-724 / LFS-834: Checkin a questionnaire when closing it

IntegralProgrammer committed:
LFS-724 / LFS-835: Verify if a form is already checked out before editing it

Only checkout a Form from JCR once its JSON data has been loaded and we are in edit mode

IntegralProgrammer committed:
LFS-724 / LFS-835: Verify if a form is already checked out before editing it

Display a colored warning message 'Another user is editing' if the jcr:isCheckedOut property is true at the moment of fetching the Form's JSON

IntegralProgrammer committed:
LFS-724 / LFS-835: Verify if a form is already checked out before editing it

Check-in a Form immediately after it is created so that the 'Another user is editing' warning message isn't displayed upon initial Form load

IntegralProgrammer committed:
LFS-724 / LFS-926: Update jcr:lastModified and jcr:lastModifiedBy when updating a mix:lastModified node
Update jcr:lastModified and jcr:lastModifiedBy Form and Questionnaire properties only upon check-in

IntegralProgrammer committed:
LFS-724 / LFS-879: Display the form's lastModified timestamp in the UI

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/Form.jsx:
```diff
@@ -110,24 +110,35 @@ function Form (props) {
       return (() => {
         // cleanup event handler
         window.removeEventListener("beforeunload", saveData);
+        window.removeEventListener("beforeunload", performCheckIn);
```
I think there's a race condition here, the checkin may finish before the save completes, which would cause the save to actually fail. I would chain these two requests.

Hm, another problem might be that in Firefox, when closing the page the request is sent, but a native confirmation dialog pops up asking if I'm sure I want to close this page. Canceling it means that now the form is checked in, but I can still edit it. We should try to get rid of that native popup, but I don't think it is a blocker for now.

sdumitriu replied:
Yep, when testing just encountered this race issue, edited the form, saved, clicked on Dashboard, got a login prompt.

The login prompt is actually bothersome, the actual problem is a 500 server error (due to the node being checked in), but the login dialog keeps popping up, and there's no way to make it go away other than refreshing the page. We need some more error handling in the fetchWithRelogin code to make sure we don't try to solve any error with another login dialog.

IntegralProgrammer replied:
More advanced handling of HTTP 500 response codes by `fetchWithReLogin` should be addressed in the future, but for now the issue of the HTTP 500 response occurring in this scenario has been fixed by the `CheckinSlingPostProcessor` which can  perform a save + checkin in one request.

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Form.jsx:
```diff
@@ -349,9 +369,17 @@ function Form (props) {
             : ""
           }
+          {
+            data && data['jcr:isCheckedOut'] ?
+            <Typography variant="overline" className={classes.warningStatus}>Another user is editing</Typography>
+            : ""
```
We should display "Another user is editing" only in edit mode.
```suggestion
            isEdit && data && data['jcr:isCheckedOut'] &&
            <Typography variant="overline" className={classes.warningStatus}>Another user is editing</Typography>
```

marta- replied:
Or we can leave it as is. As @IntegralProgrammer suggested, the warning message might dissuade users from clicking on "Edit" and thus there will be fewer conflicts.

sashaandjic commented:
List of tested sub-tasks: 
- `LFS-879 - Display the form's lastModified timestamp in the UI`
- `LFS-835 - Verify if a form is already checked out before editing it` 
- `LFS-834 -Checkin a questionnaire when closing it


Suggestion: 
- add the name of the user who is editing; 
- prevent opening the same page in multiple tabs [LFS-705](https://phenotips.atlassian.net/browse/LFS-705) 
- why don't we checkIn on save? I have a form created by Mary. Then got user Jane change it and save it. Unless Jane closes the form, the lastModified will not change, even though Mary can see Jane's edits?
- Prevent saving of unchanged data [LFS-866](https://phenotips.atlassian.net/browse/LFS-866)

_Not related to this PR: when I wanted to open the user list, I got the change password dialog (the weird problem that we cannot constantly reproduce)_

sdumitriu commented on modules/data-entry/src/main/frontend/src/dataHomepage/NewFormDialog.jsx:
```diff
@@ -79,7 +79,15 @@ function NewFormDialog(props) {
           // FIXME: Would be better to somehow obtain the router prefix from props
           // but that is not currently possible
-          props.history.push("/content.html" + URL + '.edit');
+          let checkinForm = new FormData();
```
This is not needed. Just add `:checkinNewVersionableNodes=true` to the first request that creates the form. We should also add it to the request that creates subjects and questionnaires.

An alternative is to set this globally in the configuration, meaning that all new nodes will be checkedin by default.

This is documented at https://sling.apache.org/documentation/bundles/manipulating-content-the-slingpostservlet-servlets-post.html#versionable-node-support-1

sdumitriu commented on modules/data-entry/src/main/java/ca/sickkids/ccm/lfs/DataImportServlet.java:
```diff
@@ -229,7 +210,16 @@ private void parseData(final SlingHttpServletRequest request, boolean patch) thr
         }
-        request.getResourceResolver().adaptTo(Session.class).save();
+        final Session session = request.getResourceResolver().adaptTo(Session.class);
+        session.save();
```
Something to investigate: should we save after each imported row? This may improve the import speed, since keeping too many unsaved forms and subjects in memory slows down the needed queries.

sdumitriu replied:
Something else to investigate: if we set `servlet.post.checkinNewVersionableNodes`, will we have conflicts when importing multiple forms and subjects?

marta- commented:
> Suggestions:
> 
> * add the name of the user who is editing;

We don't record the user who does checkout at this time as far as I can see. We will do it when we implement locking (separate task).

> * on save inform other users that are sitting on the same form to refresh it?

This can be a separate improvement task.

> * prevent opening the same page in multiple tabs [LFS-705](https://phenotips.atlassian.net/browse/LFS-705)
> * why don't we checkIn on save? I have a form created by Mary. Then got user Jane change it and save it. Unless Jane closes the form, the lastModified will not change, even though Mary can see Jane's edits?

This is tricky. Yes, we should update the last modified with every save or the information is inconsistent. At the same time, once you check in you'd have to check out again since you're still, technically, editing. Will this cause any issues @sdumitriu @IntegralProgrammer  ?

> * Prevent saving of unchanged data [LFS-866](https://phenotips.atlassian.net/browse/LFS-866)

sashaandjic commented:
> This is tricky. Yes, we should update the last modified with every save or the information is inconsistent. At the same time, once you check in you'd have to check out again since you're still, technically, editing. Will this cause any issues @sdumitriu @IntegralProgrammer ?

Hm, I am afraid the user might do the changes, save and just leave the form open and go home. In this case the form will not get exported.

sashaandjic commented:
I noticed that forms are looking awkward in this instance. 
e.g. ActiGraph data form (all **Display modes** are input and all **Datatypes** are text)
![image](https://user-images.githubusercontent.com/23243244/106969377-b47e8a80-6718-11eb-9b5e-18adf29ab8be.png)

marta- commented:
> I noticed that forms are looking awkward in this instance.
> e.g. ActiGraph data form (all **Display modes** are input and all **Datatypes** are text)

...Why are all the questions in all questionnaires suudenly of type text on this branch?... @sdumitriu

sdumitriu commented on modules/last-modified-property/pom.xml:
```diff
@@ -0,0 +1,90 @@
+      <groupId>org.apache.sling</groupId>
+      <artifactId>org.apache.sling.api</artifactId>
+      <version>2.11.0</version>
+      <scope>provided</scope>
+    </dependency>
+    <dependency>
+      <groupId>org.apache.sling</groupId>
+      <artifactId>org.apache.sling.scripting.jsp.taglib</artifactId>
+      <version>2.0.6</version>
+      <scope>provided</scope>
+    </dependency>
+    <dependency>
+      <groupId>org.apache.sling</groupId>
+      <artifactId>org.apache.sling.jcr.base</artifactId>
+      <version>2.2.2</version>
+      <scope>provided</scope>
+    </dependency>
+    <dependency>
+      <groupId>org.apache.jackrabbit</groupId>
+      <artifactId>jackrabbit-api</artifactId>
+      <version>2.13.1</version>
+      <scope>provided</scope>
```
These are not used.

```suggestion
```

sdumitriu commented on modules/last-modified-property/src/main/java/ca/sickkids/ccm/lfs/lastmodifiedproperty/LastModifiedEditor.java:
```diff
@@ -0,0 +1,105 @@
+import org.apache.sling.api.resource.ResourceResolver;
+
+/**
+ * A sample {@link Editor} that updates the jcr:lastModified and
```
```suggestion
 * An {@link Editor} that updates the jcr:lastModified and
```

sdumitriu commented on modules/last-modified-property/src/main/java/ca/sickkids/ccm/lfs/lastmodifiedproperty/LastModifiedEditor.java:
```diff
@@ -0,0 +1,105 @@
+/**
+ * A sample {@link Editor} that updates the jcr:lastModified and
+ * jcr:lastModifiedBy properties with the date/time and the username
+ * of the last check-in or check-out operation.
```
```suggestion
 * of the last check-in operation.
```

sdumitriu commented on modules/last-modified-property/src/main/java/ca/sickkids/ccm/lfs/lastmodifiedproperty/LastModifiedEditor.java:
```diff
@@ -0,0 +1,105 @@
+ */
+public class LastModifiedEditor extends DefaultEditor
+{
+    private static final List<String> UPDATE_TYPES = Arrays.asList("lfs/Form", "lfs/Questionnaire");
```
If possible, I would make it work for any versionable node, not just our own. `NodeState` and `NodeBuilder` don't seem to offer this information, so either get a proper Node from the session, or work with the node type definition.

sdumitriu replied:
Actually not versionable nodes, but all `mix:lastModified` instances.

sdumitriu commented on modules/last-modified-property/src/main/java/ca/sickkids/ccm/lfs/lastmodifiedproperty/LastModifiedEditor.java:
```diff
@@ -0,0 +1,105 @@
+    @Override
+    public void propertyAdded(PropertyState after) throws CommitFailedException
+    {
+        updateLastModified(after);
```
Hm, why does every newly added property change the modified date, but only when changing `isCheckedOut` to `false`? I would remove this method and only keep `propertyChanged`.

sdumitriu committed:
[misc] Apply suggestions from code review

marta- commented on modules/data-entry/src/main/java/ca/sickkids/ccm/lfs/DataImportServlet.java:
```diff
@@ -831,8 +834,8 @@ private String findSubjectId(CSVRecord row, Node typeNode)
         String result = null;
-        String[] suffices = {"", " ID"};
-        for (String suffix: suffices) {
+        String[] suffices = { "", " ID" };
+        for (String suffix : suffices) {
```
```suggestion
        String[] suffixes = { "", " ID" };
        for (String suffix : suffixes) {
```

sdumitriu commented:
> At the same time, once you check in you'd have to check out again since you're still, technically, editing. Will this cause any issues @sdumitriu @IntegralProgrammer ?

Hm, there's a very simple fix for automatic checkin and checkout, no extra code needed. If we turn `servlet.post.autoCheckout` on, then we can leave forms checked in all the time, and every save request will just checkout, save changes, checkin, with an automatic version created. However, I would not use this for the following reasons:

- we don't know that a form is edited
- if multiple users are editing and saving the same form at the same time, they will override each other's changes
- we'd have too many versions in the history
- it is not possible to do more complex version branching and tagging (we don't at the moment, but something to look into if users require it)

However, if we keep our current explicit checkout and checking, while also turning on automatic checkout/checkin, we can still have our basic locking ("someone else is editing"), but also not forbid saves while not checked out. This makes the race condition on close go away (well, not go away, just invisible), and will not cause multiple users to get errors when trying to save after one of them checks in. WDYT @marta- ?

sdumitriu committed:
[misc] Fix typo

Co-authored-by: Marta Girdea <marta.girdea@gmail.com>

marta- reviewed (changes requested):
The mysterious transformation of all questions' `dataType`s into `text` in `cardiac` and `cardiac_rehab` modes needs to be tracked down and sorted out.

sdumitriu committed:
LFS-893: Questionnaires with ConditionalGroups are broken
Fixed.
Also fixed a wrong default type for its children

sdumitriu committed:
[cleanup] Removed unused dependencies

sdumitriu committed:
LFS-897: Newly created resources should be checked in initially
LFS-898: Saving a checked in form should not fails
Turn on automatic checkin of new resources, and automatic checkout+checkin for saved resources

sdumitriu commented on modules/last-modified-property/pom.xml:
```diff
@@ -0,0 +1,71 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
I would rename this module to `versioning` (and also put the `CheckinSlingPostProcessor` in here).

IntegralProgrammer committed:
LFS-724 / LFS-926: Update jcr:lastModified and jcr:lastModifiedBy when updating a mix:lastModified node
Rename last-modified-property module to versioning

IntegralProgrammer committed:
LFS-724 / LFS-832: Checkin a form when closing it
LFS-724 / LFS-927: Any save request should also be capable of checking in the affected resource

Use a SlingPostProcessor so that Forms can optionally be checked-in upon save and thus only 1 HTTP request and not 2 need to be sent

IntegralProgrammer committed:
LFS-724 / LFS-926: Update jcr:lastModified and jcr:lastModifiedBy when updating a mix:lastModified node
Remove propertyAdded() method in LastModifiedEditor class

sdumitriu commented on modules/versioning/src/main/java/ca/sickkids/ccm/lfs/versioning/CheckinSlingPostProcessor.java:
```diff
@@ -0,0 +1,57 @@
+        } else {
+            LOGGER.warn("Running CheckinSlingPostProcessor::process(checkin=true)");
+            final Node n = request.getResource().adaptTo(Node.class);
+            n.getSession().save();
```
I don't think we need this line, the post servlet already saves everything.

```suggestion
```

IntegralProgrammer replied:
We need this line because, without it, closing an editable Form (and triggering save + check-in) will result in a HTTP 500 error with the message `Unable to perform checkin. Session has pending changes.`

sdumitriu commented on modules/versioning/src/main/java/ca/sickkids/ccm/lfs/versioning/CheckinSlingPostProcessor.java:
```diff
@@ -0,0 +1,57 @@
+            final Node n = request.getResource().adaptTo(Node.class);
+            n.getSession().save();
+            //TODO: Replace the deprecated checkin() method
+            n.checkin();
```
I think this works (blindly written):

```suggestion
            n.getSession().getWorkspace().getVersionManager().checkin(n);
```

sdumitriu replied:
This may fail if the node is not versionable. Since we are only expected to send this on versionable nodes, we should not actually trigger this error, but just to be safe, I would add a check to see if the node actually is versionable.

sdumitriu commented on modules/versioning/src/main/java/ca/sickkids/ccm/lfs/versioning/CheckinSlingPostProcessor.java:
```diff
@@ -0,0 +1,57 @@
+    @Override
+    public void process(SlingHttpServletRequest request, List<Modification> changes) throws Exception
+    {
+        RequestParameter doCheckin = request.getRequestParameter(":checkinSave");
```
```suggestion
        RequestParameter doCheckin = request.getRequestParameter(":checkin");
```

sdumitriu commented on modules/versioning/src/main/java/ca/sickkids/ccm/lfs/versioning/CheckinSlingPostProcessor.java:
```diff
@@ -0,0 +1,57 @@
+        if (doCheckin == null) {
+            LOGGER.warn("Running CheckinSlingPostProcessor::process(checkin=false)");
+        } else {
+            LOGGER.warn("Running CheckinSlingPostProcessor::process(checkin=true)");
```
```suggestion
        if (doCheckin != null) {
            LOGGER.debug("Running CheckinSlingPostProcessor::process");
```

sdumitriu commented on modules/versioning/src/main/java/ca/sickkids/ccm/lfs/versioning/CheckinSlingPostProcessor.java:
```diff
@@ -0,0 +1,57 @@
+            LOGGER.warn("Running CheckinSlingPostProcessor::process(checkin=false)");
+        } else {
+            LOGGER.warn("Running CheckinSlingPostProcessor::process(checkin=true)");
+            final Node n = request.getResource().adaptTo(Node.class);
```
Ah, this is indeed a much simpler approach than looking at modifications, we should only checkin the node that was targeted by the request.

sdumitriu commented on modules/versioning/src/main/java/ca/sickkids/ccm/lfs/versioning/LastModifiedEditorProvider.java:
```diff
@@ -0,0 +1,59 @@
+ *
+ * @version $Id$
+ */
+@Component(name = "LastModifiedEditorProvider", service = EditorProvider.class, scope = ServiceScope.SINGLETON)
```
After more practice with components, I think most of these parameters are not needed and we should be fine with just:

```suggestion
@Component
```

IntegralProgrammer committed:
LFS-724 / LFS-926: Update jcr:lastModified and jcr:lastModifiedBy when updating a mix:lastModified node
Run the updateLastModified() method from the LastModifiedEditor class on all mix:lastModified nodes

IntegralProgrammer committed:
LFS-724 / LFS-926: Update jcr:lastModified and jcr:lastModifiedBy when updating a mix:lastModified node
Update jcr:lastModified and jcr:lastModifiedBy upon each save and not just upon checkin

IntegralProgrammer committed:
LFS-724 / LFS-927: Any save request should also be capable of checking in the affected resource
Replace throwing of generic Exception with throwing of more specific javax.jcr.RepositoryException

IntegralProgrammer committed:
LFS-724 / LFS-832: Checkin a form when closing it
LFS-724 / LFS-927: Any save request should also be capable of checking in the affected resource

Rename the :checkinSave parameter to :checkin

IntegralProgrammer committed:
Revert "LFS-724: As an API consumer, I can see when a resource was last updated in any of the JSON export formats"

This reverts commit ac93a5b91f0a9fc13ba563f421108e3a838002ff.

IntegralProgrammer committed:
LFS-724 / LFS-926: Update jcr:lastModified and jcr:lastModifiedBy when updating a mix:lastModified node
LFS-914: jcr:lastModified and jcr:lastModifiedBy should not be updated if a Form is checked-in or saved without being changed.

Setting of the jcr:lastModified and jcr:lastModifiedBy properties should only be done by LastModifiedEditor and not by CheckinSlingPostProcessor

IntegralProgrammer committed:
LFS-724 / LFS-927: Any save request should also be capable of checking in the affected resource
Use the more modern (non-deprecated) Node check-in method

IntegralProgrammer committed:
LFS-724 / LFS-927: Any save request should also be capable of checking in the affected resource
Ensure that only mix:lastModified Nodes are checked in

IntegralProgrammer committed:
LFS-724 / LFS-927: Any save request should also be capable of checking in the affected resource
Clean-up the if / else code

IntegralProgrammer committed:
LFS-724 / LFS-926: Update jcr:lastModified and jcr:lastModifiedBy when updating a mix:lastModified node
Clean-up @Component annotation

IntegralProgrammer committed:
LFS-724 / LFS-832: Checkin a form when closing it

Clean-up @Component annotation

marta- reviewed (approved):
So far so good

sashaandjic reviewed (approved):
tested

sashaandjic commented:
I think I found a bug
- open any existing form (e.g. Demographic)
- initially form in not checked out - correct
- click on Edit button
- form is checked out - correct
- click on Save button
- form is saved and checked in - correct
- do some edits
- the Saved button turns into Save, but the form is not checked out - incorrect

sdumitriu commented on modules/versioning/src/main/java/ca/sickkids/ccm/lfs/versioning/CheckinSlingPostProcessor.java:
```diff
@@ -0,0 +1,56 @@
+            LOGGER.debug("Running CheckinSlingPostProcessor::process");
+            final Node n = request.getResource().adaptTo(Node.class);
+            if (n.isNodeType("mix:lastModified")) {
+                n.getSession().save();
```
```suggestion
                // At this point, all the changes are done in the session, not yet persisted, and trying to checkin the
                // node will fail. We must save first, although it is not recommended to do so in a PostProcessor.
                n.getSession().save();
```

sdumitriu committed:
[misc] Explain why we need to save

sdumitriu commented on modules/versioning/src/main/java/ca/sickkids/ccm/lfs/versioning/LastModifiedEditor.java:
```diff
@@ -0,0 +1,130 @@
+
+    private boolean isMixLastModified(NodeBuilder nodeBuilder)
+    {
+        if (nodeBuilder instanceof MemoryNodeBuilder) {
```
I don't like that we depend on an implementation detail, but it will do for the moment.

sdumitriu commented on modules/versioning/src/main/java/ca/sickkids/ccm/lfs/versioning/LastModifiedEditor.java:
```diff
@@ -0,0 +1,130 @@
+    public void propertyAdded(final PropertyState after)
+        throws CommitFailedException
+    {
+        if ("value".equals(after.getName())) {
```
This will fail to update in case "notes" are modified. Instead of only updating when the value changes, I would update except when a meaningless property changes. Is there any other than `jcr:isCheckedOut` that we should ignore? Perhaps the other versioning properties. To investigate.

sdumitriu replied:
```suggestion
        if ("value".equals(after.getName()) || "notes".equals(after.getName())) {
```

sdumitriu committed:
LFS-724 / LFS-926: Update jcr:lastModified and jcr:lastModifiedBy when updating a mix:lastModified node
Also update the modification data when changing "notes"
